pkg/querier/chunk_merge_iterator.go

```diff
@@ -153,3 +157,177 @@ type byFrom []*chunkIterator
 func (b byFrom) Len() int           { return len(b) }
 func (b byFrom) Swap(i, j int)      { b[i], b[j] = b[j], b[i] }
 func (b byFrom) Less(i, j int) bool { return b[i].From < b[j].From }
+
+type nonOverlappingIterator struct {
+	its []*chunkIterator
+
+	curr *chunkIterator
+}
+
+// newNonOverlappingIterator returns a single iterator over an slice of sorted,
+// non-overlapping iterators.
+func newNonOverlappingIterator(its []*chunkIterator) *nonOverlappingIterator {
+	return &nonOverlappingIterator{
+		its: its,
+	}
+}
+
+func (it *nonOverlappingIterator) Seek(t int64) bool {
+	for it.curr == nil || !it.curr.Seek(t) {
+		if len(it.its) == 0 {
+			return false
+		}
+
+		it.curr = it.its[0]
+		it.its = it.its[1:]
+	}
```

Comment:
I don't understand this - does this assume seeks always increase / go forward?

Would it be quicker to filter by start / finish time of the chunks, perhaps with a binary search, and then seek a single chunk?

Reply:
I particularly don't like how it mutates `it.its`.